src/Workspaces/Core/Portable/Workspace/Solution/DocumentState_LinkedFileReuse.cs

```diff
@@ -0,0 +1,208 @@
+            // to the provided source, gets the tree from it, and then wraps its root in a new tree for us.
+
+            // copy data from this entity, and pass to static helper, so we don't keep this green node alive.
+
+            var filePath = this.Attributes.SyntaxTreeFilePath;
+            var languageServices = this.LanguageServices;
+            var loadTextOptions = this.LoadTextOptions;
+            var parseOptions = this.ParseOptions;
+            var textAndVersionSource = this.TextAndVersionSource;
+            var treeSource = this.TreeSource;
+
+            var newTreeSource = GetReuseTreeSource(
+                filePath, languageServices, loadTextOptions, parseOptions, treeSource, siblingTextSource, siblingTreeSource);
+
+            return new DocumentState(
+                LanguageServices,
+                solutionServices,
+                Services,
+                Attributes,
+                _options,
+                siblingTextSource,
+                LoadTextOptions,
+                newTreeSource);
+
+            static AsyncLazy<TreeAndVersion> GetReuseTreeSource(
+                string filePath,
+                HostLanguageServices languageServices,
+                LoadTextOptions loadTextOptions,
+                ParseOptions parseOptions,
+                ValueSource<TreeAndVersion> treeSource,
+                ITextAndVersionSource siblingTextSource,
+                ValueSource<TreeAndVersion> siblingTreeSource)
+            {
+                return new AsyncLazy<TreeAndVersion>(
+                    cancellationToken => TryReuseSiblingTreeAsync(filePath, languageServices, loadTextOptions, parseOptions, treeSource, siblingTextSource, siblingTreeSource, cancellationToken),
+                    cancellationToken => TryReuseSiblingTree(filePath, languageServices, loadTextOptions, parseOptions, treeSource, siblingTextSource, siblingTreeSource, cancellationToken),
+                    cacheResult: true);
+            }
+
+            static bool TryReuseSiblingRoot(
+                string filePath,
+                HostLanguageServices languageServices,
+                LoadTextOptions loadTextOptions,
+                ParseOptions parseOptions,
+                SyntaxNode siblingRoot,
+                VersionStamp siblingVersion,
+                [NotNullWhen(true)] out TreeAndVersion? newTreeAndVersion)
+            {
+                var siblingTree = siblingRoot.SyntaxTree;
+
+                // Look for things that disqualify us from being able to use our sibling's root.
+                if (!CanReuseSiblingRoot())
+                {
+                    newTreeAndVersion = null;
+                    return false;
+                }
+
+                var treeFactory = languageServices.GetRequiredService<ISyntaxTreeFactoryService>();
+
+                var newTree = treeFactory.CreateSyntaxTree(
+                    filePath,
+                    parseOptions,
```

Comment:
Do we have a test that explicitly checks that in the case of trees that can share the same node, but have different options, that both of them have the correct options on either side? I could imagine this accidentally is the "wrong" parse options and things go badly that way.

Reply:
(I've made this mistake before...)

Reply:
i don't.  but i'll add.